src/States.cpp

```diff
@@ -0,0 +1,472 @@
+#include "States.h"
+
+#include <iostream>
+#include <stdexcept>
+#include <algorithm>
+#include <gsl/gsl_interp.h>
+#include <gsl/gsl_spline.h>
+#include <gsl/gsl_poly.h>
+#include <cmath>
+#include <float.h>
+
+namespace ale {
+  
+  /** The following helper functions are used to calculate the reduced states cache and cubic hermite 
+  to interpolate over it. They were migrated, with minor modifications, from 
+  Isis::NumericalApproximation **/
+  
+  /** Checks to see if an individual value, a, is within the doman specified by x **/
+  bool InsideDomain(const double a, std::vector<double> x) {
+    if(a + DBL_EPSILON < *min_element(x.begin(), x.end())) {
+      return false;
+    }
+    if(a - DBL_EPSILON > *max_element(x.begin(), x.end())) {
+      return false;
+    }
+    return true;
+  }
+
+  /** Find the index of the x-value in the data set that is just
+   * below the input value, a. If a is below the domain minimum,
+   * the method returns 0 as the lower index.  If a is above the
+   * domain maximum, it returns the second to last index of the
+   * data set, Size()-2, as the lower index. **/
```

Comment:
If a is below the lower end of x, then the function returns the lowest index possible (0), which makes sense to me. Why does it not return the highest index value (Size()-1) when a is above the higher end x?

Reply:
Correct me if I'm wrong: The algorithm uses the two bonding values (n, n+1). So if my data set is: {1,2,3,4,5}, and I ask for 3.2 the algorithm needs (3 and 4) aka indexes (2 and 3). If the requested values is below the min, say -10, it needs (1 and 2) aka index (0 and 1) so it can extrapolate down. If the requested values is above the max, say 9, it needs the last two (4 and 5), aka index (3 and 4) to extrapolate up.